blockchain/chain_sync/src/sync.rs

```diff
@@ -193,6 +194,38 @@ where
                             )
                             .await
                     }
+                    Some(NetworkEvent::PubsubMessage { source, message }) => {
+                        match message {
+                            forest_libp2p::PubsubMessage::Block(b) => {
+                                info!("Received block over GossipSub: {} from {}", b.header.epoch(), source.clone().unwrap());
+                                // Get bls_messages in the store or over Bitswap
+                                let mut bmsgs: Vec<UnsignedMessage> = Vec::with_capacity(b.bls_messages.len());
+                                for m in b.bls_messages.clone() {
+                                    let bmsg: UnsignedMessage = self.network.bitswap_get(m).await.unwrap();
+                                    bmsgs.push(bmsg);
+                                }
+                                // Get secp_messages in the store or over Bitswap
+                                let mut smsgs: Vec<SignedMessage> = Vec::with_capacity(b.secpk_messages.len());
+                                for m in b.secpk_messages.clone() {
+                                    let smsg: SignedMessage = self.network.bitswap_get(m).await.unwrap();
+                                    smsgs.push(smsg);
+                                }
```

Comment:
two problems I see with these
1. Unwraps when error is very common
2. you're performing the requests in sequence since you're awaiting on all of them one at a time

Either you should spawn tasks for each and join after all the requests have started, or the bitswap_get should return the receiver channel and then all should be awaited after requests sent

Reply:
Good catch. Didn't see any performance issues because i was running on devnet, but its pretty clear this will not work in mainnet